README.md
# `alto2txt`: Extract plain text from newspaper OCR scans

![GitHub](https://img.shields.io/github/license/Living-with-Machines/alto2txt) ![PyPI](https://img.shields.io/pypi/v/alto2txt) [![DOI](https://zenodo.org/badge/259340615.svg)](https://zenodo.org/badge/latestdoi/259340615)

`alto2txt` converts `XML` `ALTO`/`METS` Optical Character Recognition (OCR) scans into plaintext files with minimal metadata.

`ALTO` and `METS` are current industry standards for newspaper digitization used by hundreds of modern, large-scale newspaper digitization projects. One text file is output per article, each complemented by one `XML` metadata file. `alto2txt` converts `XML` (in `METS` `1.8`/`ALTO` `1.4`, `METS` `1.3`/`ALTO` `1.4`, `BLN` or `UKP` format) publications to plaintext articles and generates minimal metadata.

[`METS` (Metadata Encoding and Transmission Standard)](http://www.loc.gov/standards/mets/) is a standard for encoding descriptive, administrative, and structural metadata regarding objects within a digital library, expressed in `XML`. [`ALTO` (Analyzed Layout and Text Objects)](https://www.loc.gov/standards/alto/) is an [`XML` schema](https://en.wikipedia.org/wiki/XML_schema) for technical metadata describing the layout and content of text resources such as book or newspaper pages. `ALTO` is often used in combination with `METS` but can also be used independently. Details of the `ALTO` schema are avilable at https://github.com/altoxml/schema.

## [Full documentation and demo instructions.](https://living-with-machines.github.io/alto2txt/#/)

## Installation

### Installation using an Anaconda environment

We recommend installation via Anaconda:

* Refer to the [Anaconda website and follow the instructions](https://docs.anaconda.com/anaconda/install/).

* Create a new environment for `alto2txt`

```bash
conda create -n py37alto python=3.7
```

* Activate the environment:

```bash
conda activate py37alto
```

* Install `alto2txt` itself

Install `alto2txt` using pip:

```bash
pip install alto2txt
```

(For now it is still necessary to install using `pip`. In due course we plan to make `alto2txt` available through a `conda` channel, meaning that it can be installed directly using `conda` commands.)

### Installation using pip, outside an Anaconda environment

Note, the use of ``alto2txt`` outside a conda environment has not been as extensively tested as within a conda environment. Whilst we believe that this should work, please use with caution.

```bash
pip install alto2txt
```

### Installation of a test release

If you need (or want) to install a test release of `alto2txt` you will likely be advised of the specific version number to install. This examaple command will install `v0.3.1-alpha.20`:

```bash
pip install -i https://test.pypi.org/simple/ alto2txt==0.3.1a20
```

## Usage

Downsampling can be used to convert only every Nth issue of each newspaper. One text file is output per article, each complemented by one `XML` metadata file.



```
usage: alto2txt [-h] [-p [PROCESS_TYPE]] [-l [LOG_FILE]] [-d [DOWNSAMPLE]] [-n [NUM_CORES]]
                xml_in_dir txt_out_dir
alto2txt [-h] [-p [PROCESS_TYPE]] [-l [LOG_FILE]] [-d [DOWNSAMPLE]] [-n [NUM_CORES]]
         xml_in_dir txt_out_dir

Converts XML publications to plaintext articles

positional arguments:
  xml_in_dir            Input directory with XML publications
  txt_out_dir           Output directory for plaintext articles

optional arguments:
  -h, --help            show this help message and exit
  -p [PROCESS_TYPE], --process-type [PROCESS_TYPE]
                        Process type. One of: single,serial,multi,spark Default: multi
  -l [LOG_FILE], --log-file [LOG_FILE]
                        Log file. Default out.log
  -d [DOWNSAMPLE], --downsample [DOWNSAMPLE]
                        Downsample. Default 1
  -n [NUM_CORES], --num-cores [NUM_CORES]
                        Number of cores (Spark only). Default 1")
```

`xml_in_dir` is expected to hold `XML` for multiple publications, in the following structure:

```
xml_in_dir
|-- publication
|   |-- year
|   |   |-- issue
|   |   |   |-- xml_content
|   |-- year
|-- publication
```

However, if `-p|--process-type single` is provided then `xml_in_dir` is expected to hold `XML` for a single publication, in the following structure:

```
xml_in_dir
|-- year
|   |-- issue
|   |   |-- xml_content
|-- year
```

`txt_out_dir` is created with an analogous structure to `xml_in_dir`.

`PROCESS_TYPE` can be one of:

* `single`: Process single publication.
* `serial`: Process publications serially.
* `multi`: Process publications using multiprocessing (default).
* `spark`: Process publications using Spark.

`DOWNSAMPLE` must be a positive integer, default 1.

The following `XSLT` files need to be in an `extract_text.xslts` module:

* `extract_text_mets18.xslt`: `METS 1.8 XSL` file.
* `extract_text_mets13.xslt`: `METS 1.3 XSL` file.
* `extract_text_bln.xslt`: `BLN XSL` file.
* `extract_text_ukp.xslt`: `UKP XSL` file.

## Process publications

Assume folder `BNA` exists and matches the structure above.

Extract text from every publication:

```bash
alto2txt BNA txt
```

Extract text from every 100th issue of every publication:

```bash
alto2txt BNA txt -d 100
```

## Process a single publication

Extract text from every issue of a single publication:

```bash
alto2txt -p single BNA/0000151 txt
```

Extract text from every 100th issue of a single publication:

```bash
alto2txt -p single BNA/0000151 txt -d 100
```

## Configure logging

By default, logs are put in `out.log`.

To specify an alternative location for logs, use the `-l` flag e.g.

```bash
alto2txt -l mylog.txt BNA txt -d 100 2> err.log
```

## Process publications via Spark

[Information on running on spark.](spark_instructions.md)


## Future work

For a complete list of future plans see the [GitHub issues list](https://github.com/Living-with-machines/alto2txt/issues). Some highlights include:

* Export more metadata from alto, probably by parsing `mets` first.
* Check and ensure that articles that span multiple pages are pulled into a single article file.
* Smarter handling of articles spanning multiple pages.

# Copyright

## Software

Copyright 2022 The Alan Turing Institute, British Library Board, Queen Mary University of London, University of Exeter, University of East Anglia and University of Cambridge.

See [LICENSE](LICENSE) for more details.

## Example Datasets

This repo contains example datasets, which have been taken from the [British Library Research Repository](https://bl.iro.bl.uk/concern/datasets/551cdd7b-580d-472d-8efb-b7f05cf64a11) ([DOI link](https://doi.org/10.23636/1235)).

This data is "CC0 1.0 Universal Public Domain" - [No Copyright - Other Known Legal Restrictions](https://rightsstatements.org/page/NoC-OKLR/1.0/?language=en)

- There is a subset of the example data in the `demo-files` directory.
- There are adapted copies of the data in the `tests/tests/test_files` directory. These have been edited to test errors and edge cases.

# Funding and Acknowledgements

This software has been developed as part of the [Living with Machines](https://livingwithmachines.ac.uk) project.

This project, funded by the UK Research and Innovation (UKRI) Strategic Priority Fund, is a multidisciplinary collaboration delivered by the Arts and Humanities Research Council (AHRC), with The Alan Turing Institute, the British Library and the Universities of Cambridge, East Anglia, Exeter, and Queen Mary University of London.


CHANGELOG.md
# Changelog
All notable changes to this software will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).

Please maintain both this file AND the `README.md` file.

## v0.3.4

### Added
 * Added `PyPI` version and `MIT` license badges to `README.md`
 * Added `pytest-cov` with default options to assess documentation
 * Added `isort` to `.pre-commit-config.yaml` to sort import consistency
 * Added `pycln` to  `.pre-commit-config.yaml` to check unused imports
 * Added `pycln` configuration to `pyproject.toml`
 * Added `alto2txt` as a command line script in `pyproject.toml`

### Changed
 * Switch from `Apache v2.0` license to `MIT` license, inline with project recommendations.
 * Updated `mypy` in `.pre-commit-config.yaml`

### Deprecated
 * Replace `extract_publications_text.py` with the `alto2txt` `command line interface` script specified in `pyproject.toml`

### Removed
 * `setup.py`
 * `requirements.txt`

### Fixed

 * Fixed `python = ">3.6.0"` in `pyproject.toml` rather than `>3.7` for consistency with documentation
 * Fixed licensing ambiguity (now all should be `MIT`)
 * Fixed typos in `README.md`
 * Fixed surperflous imports via `pycln` in `pre-commit`

### Security

## v0.3.3

### Added

### Changed

### Deprecated

### Removed

### Fixed

### Security

## v0.3.2

### Added

### Changed

### Deprecated

### Removed

### Fixed

### Security

## v0.3.1

### Added
 * Added this changelog file.
 * Established use of [Semantic Versioning](https://semver.org/spec/v2.0.0.html) for this project.
 * Added a separate `CONTRIBUTING.md` doc for software developers notes.
 * Added a separate copyright notice for example datasets

### Changed

### Deprecated

### Removed

### Fixed

### Security

## v0.3.0 and before (prior to 2022-04-27)

For changes prior to the introduction of this changelog, see the [README.md file at that point in time](https://github.com/Living-with-machines/alto2txt/blob/54dc404ab60943c38d2e4c27a4e080cc24d4e8da/README.md).


src/alto2txt/xml.py
"""
XML utilities.
"""

import os

from lxml import etree

from alto2txt import xslts

METS_18_XSLT = "extract_text_mets18.xslt"
""" METS 1.8 XSLT """
METS_13_XSLT = "extract_text_mets13.xslt"
""" METS 1.3 XSLT """
BLN_XSLT = "extract_text_bln.xslt"
""" BLN XSLT """
UKP_XSLT = "extract_text_ukp.xslt"
""" UKP XSLT """

XSI_NS = "http://www.w3.org/2001/XMLSchema-instance"
""" XML Schema Instance namespace """
SCHEMA_LOCATION = etree.QName(XSI_NS, "schemaLocation")
""" schemaLocation element """
NO_NS_SCHEMA_LOCATION = etree.QName(XSI_NS, "noNamespaceSchemaLocation")
""" noNamespaceSchemaLocation element """

METS_NS = "http://www.loc.gov/METS/"
""" METS namespace """
METS_18_URI = "http://www.loc.gov/standards/mets/version18/mets.xsd"
""" METS 1.8 schemaLocation """
METS_13_URI = "http://schema.ccs-gmbh.com/docworks/mets-metae.xsd"
""" METS 1.3 schemaLocation """
METS_ROOT = etree.QName(METS_NS, "mets")
""" METS root element """
ALTO_ROOT = "alto"
""" ALTO root element """
BLN_ROOT = "BL_newspaper"
""" BLN root element """
BLN_PAGE_XPATH = "/BL_newspaper/BL_page"
""" XPath for BLN BL_page element """
UKP_NS = "http://tempuri.org/ncbpissue"
""" UKP namespace """
UKP_ROOT = etree.QName(UKP_NS, "UKP")
""" UKP root element """

LWM_NS = {"ukp": UKP_NS, "mets": METS_NS}
""" Namespaces of documents within Living with Machines datasets. """

XML_ROOT = "root"
""" XML metadata key. """
XML_DOCTYPE = "doctype"
""" XML metadata key. """
XML_NS = "namespaces"
""" XML metadata key. """
XML_NO_NS_SCHEMA_LOCATION = "no_ns_schema_location"
""" XML metadata key. """
XML_SCHEMA_LOCATIONS = "schema_locations"
""" XML metadata key. """

RE_METS = "(.*)[-|_](mets|METS).xml$"
""" Regular expression for METS file """


def get_path(module, *name):
    """
    Gets path to file in module, given module and relative path.

    :param module: module
    :type module: module
    :param *name: file name components
    :type *name: str
    :return: path to file
    :rtype: str
    """
    return os.path.join(os.path.dirname(module.__file__), *name)


def load_xslts():
    """
    Loads XSLTs and returns in a dictionary.

    The following XSLT files need to be in an extract_text.xslts
    module:

    * extract_text_mets18.xslt: METS 1.8 XSL file.
    * extract_text_mets13.xslt: METS 1.3 XSL file.
    * extract_text_bln.xslt: BLN XSL file.
    * extract_text_ukp.xslt: BLN UKP file.

    :return: XSLTs
    :rtype: dict(str: lxml.etree.XSLT)
    """
    xsl_transforms = {}
    for xslt_name in [METS_18_XSLT, METS_13_XSLT, BLN_XSLT, UKP_XSLT]:
        xslt_file = get_path(xslts, xslt_name)
        xsl_transforms[xslt_name] = etree.XSLT(etree.parse(xslt_file))
    return xsl_transforms


def get_xml(filename):
    """
    Gets XML document tree from file.

    :param filename: XML filename
    :type filename: str
    :return: Document tree
    :rtype: lxml.etree._ElementTree
    """
    with open(filename, "r") as f:
        document_tree = None
        parser = etree.XMLParser()
        document_tree = etree.parse(f, parser)
    return document_tree


def get_xml_metadata(document_tree):
    """
    Extracts information (root element, namespaces, schema locations,
    default schema location) from XML document tree. Returns dict of
    form:

        {
            doctype: <DOCTYPE>,
            namespaces: {<TAG>: <URL>, <TAG>: <URL>, ...},
            no_ns_schema_location: <URL> | None,
            root: <ROOT_ELEMENT>,
            schema_locations: {<URL>: <URL>, ...}
        }

    :param document_tree: Document tree
    :type document_tree: lxml.etree._ElementTree
    :return: metadata
    :rtype: dict
    """
    root_element = document_tree.getroot()
    root_element_tag = str(root_element.tag)
    doctype = str(document_tree.docinfo.doctype)
    namespaces = root_element.nsmap
    no_ns_schema_location = root_element.get(NO_NS_SCHEMA_LOCATION.text)
    schema_locations = root_element.get(SCHEMA_LOCATION.text)

    if schema_locations is not None:
        # Convert schema_locations from "namespaceURI schemaURI ..."
        # to dictionary with namespaceURI:schemaURI
        uris = schema_locations.split(" ")
        schema_locations = {uris[i]: uris[i + 1] for i in range(0, len(uris), 2)}
    else:
        schema_locations = {}
    metadata = {}
    metadata[XML_ROOT] = root_element_tag
    metadata[XML_DOCTYPE] = doctype
    metadata[XML_NS] = namespaces
    metadata[XML_NO_NS_SCHEMA_LOCATION] = no_ns_schema_location
    metadata[XML_SCHEMA_LOCATIONS] = schema_locations
    return metadata


def query_xml(document_tree, query):
    """
    Runs XPath query and returns results.

    Query string can use namespace prefixes of "ukp" and "mets"
    (as defined in LWM_NS).

    :param document_tree: Document tree
    :type document_tree: lxml.etree._ElementTree
    :param query: XPath query
    :type query: str
    :return: Query results
    :rtype: Query-specific
    """
    root_element = document_tree.getroot()
    result = root_element.xpath(query, namespaces=LWM_NS)
    return result
